src/functions.jl

```diff
@@ -6,649 +6,646 @@
-    @inbounds aux = exp( a[1] )
-    T = typeof(aux)
-    v = convert(Array{T,1}, a.coeffs)
-    coeffs = similar(v)
-    @inbounds coeffs[1] = aux
-    @inbounds for k = 1:a.order
-        coeffs[k+1] = expHomogCoef(k, v, coeffs)
-    end
-    Taylor1( coeffs, a.order )
-end
-
-# Homogeneous coefficients for exp
-doc"""
-    expHomogCoef(kcoef, ac, coeffs)
-
-Compute the `k-th` expansion coefficient of $c = \exp(a)$ given by
-
-\begin{equation*}
-c_k = \frac{1}{k} \sum_{j=0}^{k-1} (k-j) a_{k-j} c_j,
-\end{equation*}
 
-with $a$ a `Taylor1` polynomial.
-
-Inputs are the `kcoef`-th coefficient, the vector of the expansion coefficients
-`ac` of $a$, and the already calculated expansion coefficients `coeffs` of `c`.
-"""
-function expHomogCoef{T<:Number}(kcoef::Int, ac::Array{T,1}, coeffs::Array{T,1})
-    kcoef == 0 && return exp(ac[1])
-    coefhomog = zero(T)
-    @inbounds for i = 0:kcoef-1
-        coefhomog += (kcoef-i) * ac[kcoef-i+1] * coeffs[i+1]
-    end
-    coefhomog = coefhomog/kcoef
-    coefhomog
-end
+# Functions
+for T in (:Taylor1, :TaylorN)
+    @eval begin
+        ## exp ##
+        function exp(a::$T)
+            if $T == Taylor1
+                order = a.order
+                @inbounds aux = exp( a[1] )
+                c = $T( aux, order )
+            else
+                order = get_order()
+                @inbounds aux = exp( a[1][1] )
+                c = $T( aux, order )
+            end
+
+            @inbounds for k = 1:order
+                exp!(c, a, k)
+            end
+            return c
+        end
 
-## Log ## 
-doc"""
-    log(a)
+        ## log ##
+        function log(a::$T)
+            if $T == Taylor1
+                findfirst(a)>0 && throw(
+                    ArgumentError("Impossible to expand `log` around 0."))
+                order = a.order
+                @inbounds aux = log( a[1] )
+                c = $T( aux, order )
+            else
+                @inbounds a0 = a[1][1]
+                if a0 == zero(a0)
+                    throw(ArgumentError(
+                    """The 0-th order `TaylorN` coefficient must be non-zero
+                    in order to expand `log` around 0."""))
+                end
+                order = get_order()
+                @inbounds aux = log( a[1][1] )
+                c = $T( aux, order )
+            end
+
+            @inbounds for k = 1:order
+                log!(c, a, k)
+            end
+            return c
+        end
 
-Return the Taylor expansion of $\log(a)$, of order `a.order`, for `a::Taylor1`
-polynomial.
+        ## sin and cos ##
+        sin(a::$T) = sincos(a)[1]
+        cos(a::$T) = sincos(a)[2]
+        function sincos(a::$T)
+            if $T == Taylor1
+                order = a.order
+                @inbounds s = $T( sin(a[1]), order )
+                @inbounds c = $T( cos(a[1]), order )
+            else
+                order = get_order()
+                @inbounds s = $T( sin(a[1][1]), order )
```

Comment:
How about defining a function `constant_term` (or similar name) that does this correctly for Taylor1 and TaylorN.